api/http-api.md

```diff
@@ -0,0 +1,177 @@
+    there's a workflow named `general-call`.
+  * Use plural nouns where an identifier may be provided to differentiate
+    * But singular nouns are not _required_ in other places, can still be plural
+* Response codes
+  * 201 success if a creation
+  * 202 if action eventually consistent
+  * 200 other all other successes
+  * Do not use 204 for no response, even if there's no response because we may add response later, just 200 everything
+* Methods:
+  * `GET` if read-only
+    * Noun as last URL term
+    * Use query parameters for commonly set items to avoid relying on request bodies (but still rely on bodies as
+      needed). This may require some gRPC gateway customization.
+  * `PUT` if strongly consistent add/update on fixed-URL w/ idempotency
+    * Noun as last URL term
+  * `DELETE` if item is actually deleted (i.e. unreachable from `GET`)
+    * Noun as last URL term
+  * `POST` for everything else
+    * Noun as last URL term if for a create, verb if for an action
+  * `PATCH` is not needed because we don't adhere to common CRUD
+    * TODO(cretz): Needs discussion
+
+### RPCs
+
+Below is list of gRPC calls to HTTP calls so far and some notes/questions beneath each. Some notes/questions beneath one
+actually apply to many but we are not repeating ourselves too much here.
+
+* Namespaces:
+  * `RegisterNamespace` - `POST /api/v1/namespaces`
+  * `DescribeNamespace` - `GET /api/v1/namespaces/{namespace}`
+  * `ListNamespaces` - `GET /api/v1/namespaces`
+  * `UpdateNamespace` - `PUT /api/v1/namespaces/{namespace}`
+    * TODO(cretz): Confirm strongly consistent and idempotent or not
+  * `DeprecateNamespace` - `DELETE /api/v1/namespaces/{namespace}`
+    * Is this eventually consistent? Will determine response code
+    * Does this prevent returning from `GET`? Will determine HTTP method
+* Workflows:
+  * `StartWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}`
+    * May accept header as `identity` and `request_id` on these calls
+  * `GetWorkflowExecutionHistory` - `GET /api/v1/namespaces/{namespace}/workflows/{workflow_id}/history`
+    * `?nextPageToken` can be a query param even though it's big
+  * `GetWorkflowExecutionHistoryReverse` - `GET /api/v1/namespaces/{namespace}/workflows/{workflow_id}/history-reverse`
+  * `RequestCancelWorkflowExecutionRequest` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/cancel`
+    * Intentionally not taking the run ID in the URL, but should it be `?runId=` query param?
+  * `SignalWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/signal/{signal_name}`
+  * `SignalWithStartWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/signal-with-start/{signal_name}`
+  * `ResetWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/reset`
+  * `TerminateWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/terminate`
```

Comment:
I see this is a `POST`—is it not idempotent? What happens if `terminate` is called twice for the same workflow? I dug through the server code a bit and it seems like it will just fail the second time (please correct me), but the actual state of the system will not change.

(I have similar questions about why some other things are `POST`s instead of `PUT`s, but I suspect there's something I may be missing here so I'll just ask about this one for now.)

Reply:
D'oh, you're right.